Move to modern `pyproject.toml` for packaging spinup-forecast

ma595 opened the pull request:
- Introduce `pyproject.toml`
- Update tests and all import statements.
- Add NetCDF test to verify file can be loaded and use pythonpath package.
- Use `xarray[io]` instead of `xarray[complete]`. Bump pytest version so that pythonpath is compatible.

Closes #64

ma595 committed:
build(structure): restructure repo with src layout

- introduce pyproject.toml, update tests and all import statements
- fix import paths from lib to src
- add NetCDF test to verify file can be loaded and use pythonpath package
- use xarray[io] instead of xarray[complete]
- bump Pytest version so that pythonpath is compatible
- various src cleanup

ma595 committed:
test: move test data download script to tests dir

ma595 committed:
chore(lint): lint fixes to core/test files

- chore(lint): lint fixes to forecast.py
- chore(lint): update exclude paths in ruff.toml
- chore(lint): correct test_forecast.py errors
- chore(lint): Fix linting failure

ma595 committed:
refactor(restart): various style fixes

- fix restart.py so that ruff check now passes
- converted French notes to English
- cleaned up docstrings
- annotated unused variables
- fixed comments

We need to review or update:
- the comment referencing Isaac
- the update_e3t function
- static type hinting
- some comments seem out of date / wrong
- no testing of any of the functions

isaacaka commented on ruff.toml:
```diff
@@ -24,7 +24,7 @@ extend-select = ["D417"]
 ignore = ["SIM108", "RUF002", "PLR0913", "PLR2004", "D100", "D104", "D105"]
 
-exclude = ["standalone/**", "lib/density.py", "lib/optimization.py", "lib/restart.py", "main_restart.py", "Notebooks/Restart.ipynb", "Notebooks/Jumper.ipynb"]
+exclude = ["standalone/**", "src/nemo_spinup_forecast/optimization.py", "main_restart.py", "Notebooks/Restart.ipynb", "Notebooks/Jumper.ipynb", "src/nemo_spinup_forecast/optimization.py"]
```
` "src/nemo_spinup_forecast/optimization.py"`
The above is included twice

isaacaka commented on pyproject.toml:
```diff
@@ -0,0 +1,46 @@
+]
+pythonpath = ["src"]
+
+addopts = "-ra -q"
```
```suggestion
addopts = "-ra"
```

I think removing '-q' so that the user knows how many tests will run and generally outputs the tests results in a nice format

ma595 replied:
I agree, done.

ma595 committed:
build: update pyproject dependencies

- add pre-commit as dev dependency
- add lower bounds on main and dev dependencies
- remove '-q' so that the user knows how many tests will run and generally outputs the tests results in a nicer format.
- bump to Python==3.10

Co-authored-by: isaacaka <ia464@cam.ac.uk>

isaacaka commented on main.py:
```diff
@@ -6,10 +6,12 @@
 import numpy as np
 
 sys.path.insert(0, "./lib/")
```
`sys.path.insert(0, "./lib/")`
Can be removed

isaacaka reviewed (commented):
All tests pass. 
In order to run `main.py`
Need to install in package in edit mode `pip install -e .` or src to be added to the pythonpath via: PYTHONPATH=src python main.py --ye True --start 20 --end 50 --comp 1 --steps 30 --path /path/to/data

What do you think is the best approach for this?

isaacaka reviewed (approved):
Good to merge

ma595 committed:
feat(cli): add module execution

- update docs for module execution
- amend README.md

ma595 committed:
feat(config): move config files and add cli flags

The behaviour now is that the default ocean_terms.DINO.yaml and
techniques_config.yaml will always be found, ass they are provided
as part of the package. Users can easily override by providing command
line arguments:

- Enables the path resolution to succeed
- Move ocean_terms.yaml and techniques_config.yaml
- Introduced CLI flags --ocean-terms and --techniques-config
- Simplified technique loaders to accept direct file paths

test: update paths to src and point to configs

- update paths from lib to src
- point to package techniques_config.yaml

docs(readme): add cli docs to pipeline description

isaacaka commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+version = "0.1.0"
+description = "Hybrid machine learning tooling for NEMO spinup optimization"
+readme = "README.md"
+requires-python = ">=3.9"
```
Python 3.9 has reached end of life so we can bump up our minimum supported version to 3.10

isaacaka commented on src/nemo_spinup_forecast/cli.py:
```diff
@@ -0,0 +1,234 @@
+    comp,
+    dr_technique,
+    forecast_technique,
+    ocean_terms_path: Path | str | None = None,
```
Typing syntax used requires python 3.10>. Breaks support for python 3.9. Python 3.9 has reached end of life so we can bump up our minimum supported version to 3.10

isaacaka commented on src/nemo_spinup_forecast/utils.py:
```diff
@@ -100,43 +101,52 @@ def prepare(term, filename, simu_path, start, end, ye, comp, dr_technique):
 
 
-def get_ocean_term(ocean_property):
+def get_ocean_term(ocean_property, yaml_path: Path | str | None = None):
```
Typing syntax used requires python 3.10>. Breaks support for python 3.9. Python 3.9 has reached end of life so we can bump up our minimum supported version to 3.10

isaacaka commented on README.md:
````diff
@@ -66,32 +92,34 @@ Terms:
-      --ye True \
-      --start 20 --end 50 \
-      --comp 0.9 --steps 20
-    ```
````
Update README instructions to include config file arguments

ma595 replied:
I think this is already included?

isaacaka replied:
Only in the quick start part, not in the full instructions under heading `8. End‑to‑End Steps for Running Spin‑Up NEMO`

ma595 replied:
fixed

isaacaka reviewed (changes requested):
Update README instructions to include additional config file arguments
Python 3.9 has reached end of life so we can bump up our minimum supported version to 3.10

Forecasting pipeline run successfully
All tests pass

isaacaka reviewed (approved):
Nice
